agent/hcp/telemetry/otel_sink.go

```diff
@@ -0,0 +1,189 @@
+	}
+
+	return &OTELSink{
+		spaceReplacer:        strings.NewReplacer(" ", "_"),
+		logger:               opts.Logger.Named("otel_sink"),
+		meterProvider:        meterProvider,
+		meter:                &meter,
+		mutex:                sync.Mutex{},
+		gaugeStore:           gs,
+		gaugeInstruments:     make(map[string]*instrument.Float64ObservableGauge, 0),
+		counterInstruments:   make(map[string]*instrument.Float64Counter, 0),
+		histogramInstruments: make(map[string]*instrument.Float64Histogram, 0),
+	}, nil
+}
+
+// SetGauge emits a Consul gauge metric.
+func (o *OTELSink) SetGauge(key []string, val float32) {
+	o.SetGaugeWithLabels(key, val, nil)
+}
+
+// AddSample emits a Consul histogram metric.
+func (o *OTELSink) AddSample(key []string, val float32) {
+	o.AddSampleWithLabels(key, val, nil)
+}
+
+// IncrCounter emits a Consul counter metric.
+func (o *OTELSink) IncrCounter(key []string, val float32) {
+	o.IncrCounterWithLabels(key, val, nil)
+}
+
+// AddSampleWithLabels emits a Consul gauge metric that gets
+// registed by an OpenTelemetry Histogram instrument.
+func (o *OTELSink) SetGaugeWithLabels(key []string, val float32, labels []gometrics.Label) {
+	k := o.flattenKey(key)
+
+	// Set value in global Gauge store.
+	o.gaugeStore.Store(k, float64(val), toAttributes(labels))
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	// If instrument does not exist, create it and register callback to emit last value in global Gauge store.
+	if _, ok := o.gaugeInstruments[k]; !ok {
+		inst, err := (*o.meter).Float64ObservableGauge(k, instrument.WithFloat64Callback(o.gaugeStore.gaugeCallback(k)))
+		if err != nil {
+			o.logger.Error("Failed to emit gauge: %w", err)
+			return
+		}
+		o.gaugeInstruments[k] = &inst
+	}
+}
+
+// AddSampleWithLabels emits a Consul sample metric that gets registed by an OpenTelemetry Histogram instrument.
+func (o *OTELSink) AddSampleWithLabels(key []string, val float32, labels []gometrics.Label) {
+	k := o.flattenKey(key)
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	inst, ok := o.histogramInstruments[k]
+	if !ok {
+		histogram, err := (*o.meter).Float64Histogram(k)
+		if err != nil {
+			o.logger.Error("Failed to emit gauge: %w", err)
+			return
+		}
+		inst = &histogram
+		o.histogramInstruments[k] = inst
+	}
+
+	attrs := toAttributes(labels)
+	(*inst).Record(context.TODO(), float64(val), attrs...)
+}
+
+// IncrCounterWithLabels emits a Consul counter metric that gets registed by an OpenTelemetry Histogram instrument.
+func (o *OTELSink) IncrCounterWithLabels(key []string, val float32, labels []gometrics.Label) {
+	k := o.flattenKey(key)
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	inst, ok := o.counterInstruments[k]
+	if !ok {
+		counter, err := (*o.meter).Float64Counter(k)
+		if err != nil {
+			o.logger.Error("Failed to emit gauge: %w", err)
+			return
+		}
+
+		inst = &counter
+		o.counterInstruments[k] = inst
+	}
+
+	attrs := toAttributes(labels)
+	(*inst).Add(context.TODO(), float64(val), attrs...)
```

Comment:
Should we be using a cancelable context here?

Reply:
same as https://github.com/hashicorp/consul/pull/17159#discussion_r1191361841

Reply:
Resolving this and we can discuss in that thread!